iroh/src/magicsock/transports/relay.rs

```diff
@@ -100,6 +100,25 @@ impl RelayTransport {
                 }
             };
 
+            if buf_out.len() < dm.datagrams.contents.len() {
+                // Our receive buffer isn't big enough to process this datagram.
+                // Continuing would cause a panic.
+                warn!(
+                    quinn_buf_len = buf_out.len(),
+                    datagram_len = dm.datagrams.contents.len(),
+                    segment_size = ?dm.datagrams.segment_size,
+                    "dropping received datagram: quinn buffer too small"
+                );
+                break;
```

Comment:
One, in my opinion large, downside of your approach is that you have to plumb the signal of how large this an be all the way through to the ActiveRelayActor.  I find all that plumbing really unfortunate and complex.

Have you considered having an `Option<Datagrams>` on the `RelayTransport` itself?  I think that way you could store the split off ones here locally.  When you get called next you first consume that and then continue reading off the channel.  Would that not be a lot simpler?

Reply:
Didn't consider that.
But thinking about how I'd do this, I'm somewhat afraid of messing up the waking logic in that case.
I'd prefer not to have to mess with wakers :S

Reply:
Do you really need to mess with wakers?  When you receive something that doesn't fit you will return a `Poll::Ready` and no waker is expected to be installed.  Next time it gets called it again returns `Poll::Ready` because there's something in the Option.  When the Option is empty you poll the channel and things behave as before.